### ISSUE TITLE
Applying/removing tags on IMAP accounts results in duplicate messages

### ISSUE THREAD

**Issue by thomasst** [nylas/sync-engine#182]
IMAP account on a dovecot server (IDLE/CONDSTORE disabled).

Thread with three messages:

```
% curl 'http://127.0.0.1:5555/n/copo7v49qoctigs5iu18y133k/threads?view=expanded'
[
    {
        "drafts": [],
        "first_message_timestamp": 1438722610,
        "folders": [
            {
                "display_name": "INBOX",
                "id": "1913losem1u6cmd4b8iuke9tu",
                "name": "inbox"
            },
            {
                "display_name": "Sent",
                "id": "d8b1yvg9rfip2vh4ncwe8gui8",
                "name": "sent"
            }
        ],
        "has_attachments": false,
        "id": "4lfikk5147ik6gwmp4obh4sfe",
        "last_message_received_timestamp": 1438722611,
        "last_message_timestamp": 1438722612,
        "messages": [
            {
                "bcc": [],
                "cc": [],
                "date": 1438722610,
                "files": [],
                "folder": {
                    "display_name": "INBOX",
                    "id": "1913losem1u6cmd4b8iuke9tu",
                    "name": "inbox"
                },
                "from": [
                    {
                        "email": "john@example.com",
                        "name": ""
                    }
                ],
                "id": "e1em9yn5nji4moagqodusdkdu",
                "message_id_header": "<m1@example.com>",
                "namespace_id": "copo7v49qoctigs5iu18y133k",
                "object": "message",
                "reply_to": [
                    {
                        "email": "replyto@example.com",
                        "name": ""
                    }
                ],
                "snippet": "first message",
                "starred": false,
                "subject": "Hello",
                "thread_id": "4lfikk5147ik6gwmp4obh4sfe",
                "to": [
                    {
                        "email": "tester@example.com",
                        "name": ""
                    }
                ],
                "unread": false
            },
            {
                "bcc": [],
                "cc": [
                    {
                        "email": "cc@example.com",
                        "name": ""
                    }
                ],
                "date": 1438722611,
                "files": [],
                "folder": {
                    "display_name": "INBOX",
                    "id": "1913losem1u6cmd4b8iuke9tu",
                    "name": "inbox"
                },
                "from": [
                    {
                        "email": "john@example.com",
                        "name": ""
                    }
                ],
                "id": "26gqtkz8zg1imuyt5f3mv9f20",
                "message_id_header": "<m2@example.com>",
                "namespace_id": "copo7v49qoctigs5iu18y133k",
                "object": "message",
                "reply_to": [],
                "snippet": "second message",
                "starred": false,
                "subject": "Hello",
                "thread_id": "4lfikk5147ik6gwmp4obh4sfe",
                "to": [
                    {
                        "email": "tester@example.com",
                        "name": ""
                    }
                ],
                "unread": true
            },
            {
                "bcc": [],
                "cc": [],
                "date": 1438722612,
                "files": [],
                "folder": {
                    "display_name": "INBOX",
                    "id": "1913losem1u6cmd4b8iuke9tu",
                    "name": "inbox"
                },
                "from": [
                    {
                        "email": "tester@example.com",
                        "name": ""
                    }
                ],
                "id": "6mfobbyln7sfkhkygei789efa",
                "message_id_header": "<m3@example.com>",
                "namespace_id": "copo7v49qoctigs5iu18y133k",
                "object": "message",
                "reply_to": [],
                "snippet": "third message",
                "starred": false,
                "subject": "Re: Hello",
                "thread_id": "4lfikk5147ik6gwmp4obh4sfe",
                "to": [
                    {
                        "email": "john@example.com",
                        "name": ""
                    }
                ],
                "unread": false
            }
        ],
        "namespace_id": "copo7v49qoctigs5iu18y133k",
        "object": "thread",
        "participants": [
            {
                "email": "john@example.com",
                "name": ""
            },
            {
                "email": "cc@example.com",
                "name": ""
            },
            {
                "email": "tester@example.com",
                "name": ""
            }
        ],
        "snippet": "third message",
        "starred": false,
        "subject": "Hello",
        "tags": [
            {
                "id": "inbox",
                "name": "INBOX"
            },
            {
                "id": "sent",
                "name": "Sent"
            },
            {
                "id": "unread",
                "name": "unread"
            }
        ],
        "unread": true,
        "version": 2
    }
]
```

Removing inbox tag:
```
% curl -XPUT 'http://127.0.0.1:5555/n/copo7v49qoctigs5iu18y133k/threads/4lfikk5147ik6gwmp4obh4sfe' -d '{"remove_tags": "inbox"}'
{
    "draft_ids": [],
    "first_message_timestamp": 1438722610,
    "folders": [
        {
            "display_name": "Archive",
            "id": "4q5bo9ynswnnzcwlff54xd27o",
            "name": "archive"
        }
    ],
    "has_attachments": false,
    "id": "4lfikk5147ik6gwmp4obh4sfe",
    "last_message_received_timestamp": 1438722611,
    "last_message_timestamp": 1438722612,
    "message_ids": [
        "e1em9yn5nji4moagqodusdkdu",
        "26gqtkz8zg1imuyt5f3mv9f20",
        "6mfobbyln7sfkhkygei789efa"
    ],
    "namespace_id": "copo7v49qoctigs5iu18y133k",
    "object": "thread",
    "participants": [
        {
            "email": "john@example.com",
            "name": ""
        },
        {
            "email": "cc@example.com",
            "name": ""
        },
        {
            "email": "tester@example.com",
            "name": ""
        }
    ],
    "snippet": "third message",
    "starred": false,
    "subject": "Hello",
    "tags": [
        {
            "id": "archive",
            "name": "Archive"
        },
        {
            "id": "unread",
            "name": "unread"
        }
    ],
    "unread": true,
    "version": 5
}
```

Fetching the thread right after removing the inbox tag looks correct:
```
% curl 'http://127.0.0.1:5555/n/copo7v49qoctigs5iu18y133k/threads?view=expanded'
[
    {
        "drafts": [],
        "first_message_timestamp": 1438722610,
        "folders": [
            {
                "display_name": "Archive",
                "id": "4q5bo9ynswnnzcwlff54xd27o",
                "name": "archive"
            }
        ],
        "has_attachments": false,
        "id": "4lfikk5147ik6gwmp4obh4sfe",
        "last_message_received_timestamp": 1438722611,
        "last_message_timestamp": 1438722612,
        "messages": [
            {
                "bcc": [],
                "cc": [],
                "date": 1438722610,
                "files": [],
                "folder": {
                    "display_name": "Archive",
                    "id": "4q5bo9ynswnnzcwlff54xd27o",
                    "name": "archive"
                },
                "from": [
                    {
                        "email": "john@example.com",
                        "name": ""
                    }
                ],
                "id": "e1em9yn5nji4moagqodusdkdu",
                "message_id_header": "<m1@example.com>",
                "namespace_id": "copo7v49qoctigs5iu18y133k",
                "object": "message",
                "reply_to": [
                    {
                        "email": "replyto@example.com",
                        "name": ""
                    }
                ],
                "snippet": "first message",
                "starred": false,
                "subject": "Hello",
                "thread_id": "4lfikk5147ik6gwmp4obh4sfe",
                "to": [
                    {
                        "email": "tester@example.com",
                        "name": ""
                    }
                ],
                "unread": false
            },
            {
                "bcc": [],
                "cc": [
                    {
                        "email": "cc@example.com",
                        "name": ""
                    }
                ],
                "date": 1438722611,
                "files": [],
                "folder": {
                    "display_name": "Archive",
                    "id": "4q5bo9ynswnnzcwlff54xd27o",
                    "name": "archive"
                },
                "from": [
                    {
                        "email": "john@example.com",
                        "name": ""
                    }
                ],
                "id": "26gqtkz8zg1imuyt5f3mv9f20",
                "message_id_header": "<m2@example.com>",
                "namespace_id": "copo7v49qoctigs5iu18y133k",
                "object": "message",
                "reply_to": [],
                "snippet": "second message",
                "starred": false,
                "subject": "Hello",
                "thread_id": "4lfikk5147ik6gwmp4obh4sfe",
                "to": [
                    {
                        "email": "tester@example.com",
                        "name": ""
                    }
                ],
                "unread": true
            },
            {
                "bcc": [],
                "cc": [],
                "date": 1438722612,
                "files": [],
                "folder": {
                    "display_name": "Archive",
                    "id": "4q5bo9ynswnnzcwlff54xd27o",
                    "name": "archive"
                },
                "from": [
                    {
                        "email": "tester@example.com",
                        "name": ""
                    }
                ],
                "id": "6mfobbyln7sfkhkygei789efa",
                "message_id_header": "<m3@example.com>",
                "namespace_id": "copo7v49qoctigs5iu18y133k",
                "object": "message",
                "reply_to": [],
                "snippet": "third message",
                "starred": false,
                "subject": "Re: Hello",
                "thread_id": "4lfikk5147ik6gwmp4obh4sfe",
                "to": [
                    {
                        "email": "john@example.com",
                        "name": ""
                    }
                ],
                "unread": false
            }
        ],
        "namespace_id": "copo7v49qoctigs5iu18y133k",
        "object": "thread",
        "participants": [
            {
                "email": "john@example.com",
                "name": ""
            },
            {
                "email": "cc@example.com",
                "name": ""
            },
            {
                "email": "tester@example.com",
                "name": ""
            }
        ],
        "snippet": "third message",
        "starred": false,
        "subject": "Hello",
        "tags": [
            {
                "id": "archive",
                "name": "Archive"
            },
            {
                "id": "unread",
                "name": "unread"
            }
        ],
        "unread": true,
        "version": 5
    }
]
```

However, after waiting for a few seconds, we can see duplicate messages in the thread. In this case, the first two messages are duplicated. In some cases, there are even 4 copies of a message.
```
% curl 'http://127.0.0.1:5555/n/copo7v49qoctigs5iu18y133k/threads?view=expanded'
[
    {
        "drafts": [],
        "first_message_timestamp": 1438722610,
        "folders": [
            {
                "display_name": "Archive",
                "id": "4q5bo9ynswnnzcwlff54xd27o",
                "name": "archive"
            }
        ],
        "has_attachments": false,
        "id": "4lfikk5147ik6gwmp4obh4sfe",
        "last_message_received_timestamp": 1438722612,
        "last_message_timestamp": 1438722612,
        "messages": [
            {
                "bcc": [],
                "cc": [],
                "date": 1438722610,
                "files": [],
                "folder": {
                    "display_name": "Archive",
                    "id": "4q5bo9ynswnnzcwlff54xd27o",
                    "name": "archive"
                },
                "from": [
                    {
                        "email": "john@example.com",
                        "name": ""
                    }
                ],
                "id": "e1em9yn5nji4moagqodusdkdu",
                "message_id_header": "<m1@example.com>",
                "namespace_id": "copo7v49qoctigs5iu18y133k",
                "object": "message",
                "reply_to": [
                    {
                        "email": "replyto@example.com",
                        "name": ""
                    }
                ],
                "snippet": "first message",
                "starred": false,
                "subject": "Hello",
                "thread_id": "4lfikk5147ik6gwmp4obh4sfe",
                "to": [
                    {
                        "email": "tester@example.com",
                        "name": ""
                    }
                ],
                "unread": true
            },
            {
                "bcc": [],
                "cc": [],
                "date": 1438722610,
                "files": [],
                "folder": {
                    "display_name": "Archive",
                    "id": "4q5bo9ynswnnzcwlff54xd27o",
                    "name": "archive"
                },
                "from": [
                    {
                        "email": "john@example.com",
                        "name": ""
                    }
                ],
                "id": "bfojn55ckq1t3er5746nkmx0z",
                "message_id_header": "<m1@example.com>",
                "namespace_id": "copo7v49qoctigs5iu18y133k",
                "object": "message",
                "reply_to": [
                    {
                        "email": "replyto@example.com",
                        "name": ""
                    }
                ],
                "snippet": "first message",
                "starred": false,
                "subject": "Hello",
                "thread_id": "4lfikk5147ik6gwmp4obh4sfe",
                "to": [
                    {
                        "email": "tester@example.com",
                        "name": ""
                    }
                ],
                "unread": false
            },
            {
                "bcc": [],
                "cc": [
                    {
                        "email": "cc@example.com",
                        "name": ""
                    }
                ],
                "date": 1438722611,
                "files": [],
                "folder": {
                    "display_name": "Archive",
                    "id": "4q5bo9ynswnnzcwlff54xd27o",
                    "name": "archive"
                },
                "from": [
                    {
                        "email": "john@example.com",
                        "name": ""
                    }
                ],
                "id": "26gqtkz8zg1imuyt5f3mv9f20",
                "message_id_header": "<m2@example.com>",
                "namespace_id": "copo7v49qoctigs5iu18y133k",
                "object": "message",
                "reply_to": [],
                "snippet": "second message",
                "starred": false,
                "subject": "Hello",
                "thread_id": "4lfikk5147ik6gwmp4obh4sfe",
                "to": [
                    {
                        "email": "tester@example.com",
                        "name": ""
                    }
                ],
                "unread": true
            },
            {
                "bcc": [],
                "cc": [
                    {
                        "email": "cc@example.com",
                        "name": ""
                    }
                ],
                "date": 1438722611,
                "files": [],
                "folder": {
                    "display_name": "Archive",
                    "id": "4q5bo9ynswnnzcwlff54xd27o",
                    "name": "archive"
                },
                "from": [
                    {
                        "email": "john@example.com",
                        "name": ""
                    }
                ],
                "id": "52lm2bs9w7kkd6m5pjytiesi3",
                "message_id_header": "<m2@example.com>",
                "namespace_id": "copo7v49qoctigs5iu18y133k",
                "object": "message",
                "reply_to": [],
                "snippet": "second message",
                "starred": false,
                "subject": "Hello",
                "thread_id": "4lfikk5147ik6gwmp4obh4sfe",
                "to": [
                    {
                        "email": "tester@example.com",
                        "name": ""
                    }
                ],
                "unread": true
            },
            {
                "bcc": [],
                "cc": [],
                "date": 1438722612,
                "files": [],
                "folder": {
                    "display_name": "Archive",
                    "id": "4q5bo9ynswnnzcwlff54xd27o",
                    "name": "archive"
                },
                "from": [
                    {
                        "email": "tester@example.com",
                        "name": ""
                    }
                ],
                "id": "6mfobbyln7sfkhkygei789efa",
                "message_id_header": "<m3@example.com>",
                "namespace_id": "copo7v49qoctigs5iu18y133k",
                "object": "message",
                "reply_to": [],
                "snippet": "third message",
                "starred": false,
                "subject": "Re: Hello",
                "thread_id": "4lfikk5147ik6gwmp4obh4sfe",
                "to": [
                    {
                        "email": "john@example.com",
                        "name": ""
                    }
                ],
                "unread": true
            },
            {
                "bcc": [],
                "cc": [],
                "date": 1438722612,
                "files": [],
                "folder": {
                    "display_name": "Archive",
                    "id": "4q5bo9ynswnnzcwlff54xd27o",
                    "name": "archive"
                },
                "from": [
                    {
                        "email": "tester@example.com",
                        "name": ""
                    }
                ],
                "id": "99en85c1n1s01j2rala7py8f1",
                "message_id_header": "<m3@example.com>",
                "namespace_id": "copo7v49qoctigs5iu18y133k",
                "object": "message",
                "reply_to": [],
                "snippet": "third message",
                "starred": false,
                "subject": "Re: Hello",
                "thread_id": "4lfikk5147ik6gwmp4obh4sfe",
                "to": [
                    {
                        "email": "john@example.com",
                        "name": ""
                    }
                ],
                "unread": false
            }
        ],
        "namespace_id": "copo7v49qoctigs5iu18y133k",
        "object": "thread",
        "participants": [
            {
                "email": "john@example.com",
                "name": ""
            },
            {
                "email": "cc@example.com",
                "name": ""
            },
            {
                "email": "tester@example.com",
                "name": ""
            }
        ],
        "snippet": "third message",
        "starred": false,
        "subject": "Hello",
        "tags": [
            {
                "id": "archive",
                "name": "Archive"
            },
            {
                "id": "unread",
                "name": "unread"
            }
        ],
        "unread": true,
        "version": 13
    }
]
```

After waiting for several minutes, the thread is back to normal:
```
% curl 'http://127.0.0.1:5555/n/copo7v49qoctigs5iu18y133k/threads?view=expanded'
[
    {
        "drafts": [],
        "first_message_timestamp": 1438722610,
        "folders": [
            {
                "display_name": "Archive",
                "id": "4q5bo9ynswnnzcwlff54xd27o",
                "name": "archive"
            }
        ],
        "has_attachments": false,
        "id": "4lfikk5147ik6gwmp4obh4sfe",
        "last_message_received_timestamp": 1438722612,
        "last_message_timestamp": 1438722612,
        "messages": [
            {
                "bcc": [],
                "cc": [],
                "date": 1438722610,
                "files": [],
                "folder": {
                    "display_name": "Archive",
                    "id": "4q5bo9ynswnnzcwlff54xd27o",
                    "name": "archive"
                },
                "from": [
                    {
                        "email": "john@example.com",
                        "name": ""
                    }
                ],
                "id": "bfojn55ckq1t3er5746nkmx0z",
                "message_id_header": "<m1@example.com>",
                "namespace_id": "copo7v49qoctigs5iu18y133k",
                "object": "message",
                "reply_to": [
                    {
                        "email": "replyto@example.com",
                        "name": ""
                    }
                ],
                "snippet": "first message",
                "starred": false,
                "subject": "Hello",
                "thread_id": "4lfikk5147ik6gwmp4obh4sfe",
                "to": [
                    {
                        "email": "tester@example.com",
                        "name": ""
                    }
                ],
                "unread": false
            },
            {
                "bcc": [],
                "cc": [
                    {
                        "email": "cc@example.com",
                        "name": ""
                    }
                ],
                "date": 1438722611,
                "files": [],
                "folder": {
                    "display_name": "Archive",
                    "id": "4q5bo9ynswnnzcwlff54xd27o",
                    "name": "archive"
                },
                "from": [
                    {
                        "email": "john@example.com",
                        "name": ""
                    }
                ],
                "id": "52lm2bs9w7kkd6m5pjytiesi3",
                "message_id_header": "<m2@example.com>",
                "namespace_id": "copo7v49qoctigs5iu18y133k",
                "object": "message",
                "reply_to": [],
                "snippet": "second message",
                "starred": false,
                "subject": "Hello",
                "thread_id": "4lfikk5147ik6gwmp4obh4sfe",
                "to": [
                    {
                        "email": "tester@example.com",
                        "name": ""
                    }
                ],
                "unread": true
            },
            {
                "bcc": [],
                "cc": [],
                "date": 1438722612,
                "files": [],
                "folder": {
                    "display_name": "Archive",
                    "id": "4q5bo9ynswnnzcwlff54xd27o",
                    "name": "archive"
                },
                "from": [
                    {
                        "email": "tester@example.com",
                        "name": ""
                    }
                ],
                "id": "99en85c1n1s01j2rala7py8f1",
                "message_id_header": "<m3@example.com>",
                "namespace_id": "copo7v49qoctigs5iu18y133k",
                "object": "message",
                "reply_to": [],
                "snippet": "third message",
                "starred": false,
                "subject": "Re: Hello",
                "thread_id": "4lfikk5147ik6gwmp4obh4sfe",
                "to": [
                    {
                        "email": "john@example.com",
                        "name": ""
                    }
                ],
                "unread": false
            }
        ],
        "namespace_id": "copo7v49qoctigs5iu18y133k",
        "object": "thread",
        "participants": [
            {
                "email": "john@example.com",
                "name": ""
            },
            {
                "email": "cc@example.com",
                "name": ""
            },
            {
                "email": "tester@example.com",
                "name": ""
            }
        ],
        "snippet": "third message",
        "starred": false,
        "subject": "Hello",
        "tags": [
            {
                "id": "archive",
                "name": "Archive"
            },
            {
                "id": "unread",
                "name": "unread"
            }
        ],
        "unread": true,
        "version": 19
    }
]
```

**Comment by thomasst** [comment]
This is still an issue and seems unrelated to https://github.com/nylas/sync-engine/issues/188

**Comment by DakotaNelson** [comment]
This should be fixed by d715f4c.

Feel free to reopen if this issue crops up again!
